### ISSUE TITLE
Using ```grad``` on ```vmap``` on ```map``` on function containing ```sinc``` results in error

### ISSUE THREAD

**Issue by Ph03n1xdust** [google/jax#10750]
Hi,

I ran into an error while trying to take the ```grad``` of a ```vmap``` on a ```map``` on a function, which contains ```sinc```.  The following code reproduces the error:

```
import jax
import jax.numpy as jnp

N_batch = 5
N_elem = 3

batch_data = jnp.reshape(jnp.arange(N_batch*N_elem),(N_batch,N_elem))

def fail(x):
    return jnp.sinc(x)

def mapfail(x):
    return jax.lax.map(fail, x)

vmapmapfail = jax.vmap(mapfail)

def loss(param):
    pos = param*batch_data
    desc = vmapmapfail(pos)
    return jnp.mean(desc)

jax.value_and_grad(loss)(0.1)
```

- Error: ```JaxStackTraceBeforeTransformation: TypeError: broadcast_in_dim broadcast_dimensions must have length equal to operand ndim; got broadcast_dimensions () for operand ndim 1.```
- Jax version: 0.3.13

I suspect this is related to the custom derivatives of the ```_sinc_maclaurin```, since commenting out both ```@partial(custom_jvp, nondiff_argnums=(0,))``` and ```@_sinc_maclaurin.defjvp``` in the jax source "solves" the problem. It is also specific to ```map```, since using a vmap instead results in no error. 

Ps.: A colleague of mine also reminded me, that the documentation of sinc, does not describe the x=0 behaviour correctly.

**Comment by jakevdp** [comment]
Thanks for the report. I'll look into the issue – Can you say more about the issue you see in the `sinc` documentation?

**Comment by sousaw** [comment]
I am afraid I am the nitpicking colleague (unexpectedly) mentioned in the report :-). My comment was actually more closely related to the part of the docstring inherited from numpy, specifically the sentence "the sinc function is sin(pi*x)/(pi*x)". Since the analytic character of sinc is important both for its applications and for the actual implementation at x=0, I would suggest defining it, for instance, as the limit when t->x of sin(pi*t)/(pi*t).

Thank you for your work on this fantastic project!

**Comment by jakevdp** [comment]
Thanks - as you said, that part of the docstring is drawn directly from `numpy.sinc`, so the tweak would have to be contributed to the numpy package itself.

**Comment by mattjj** [comment]
This is a fantastic bug! Thanks for raising it. (And thanks for the tip about the docstring too.)

I think the best fix may be to add a `full_like` primitive. (Alternatives include: defensively batching more outputs of `custom_jvp` functions, or probing custom JVP rules' vmap data dependence, but neither of those seem as good.)

A quick fix, which is basically the same thing, is to replace [these lines](https://github.com/google/jax/blob/6110be40dc66f0de9d0e3770645a7db81434910f/jax/_src/numpy/ufuncs.py#L703-L706) with

```python
if k % 2:
  return x * 0
else:
  return x * 0 + lax.full_like(x, (-1) ** (k // 2) / (k + 1))
```

The issue has to do with data dependence, and an assumption baked into `custom_jvp` about how the input-output data dependence of a custom JVP rule relates to the input-output data dependence of the function for which it is the rule. That is, if we have `f = custom_jvp(f_orig)` and `f.defjvp(f_jvp)`, our batching rule for `custom_jvp` roughly assumes that for any `in_axes: Tuple[Optional[int], ...]` we have `vmap_out_axes(f_orig, in_axes) == vmap_out_axes(lambda xs: jax.jvp(f, xs, xs), in_axes))`, which in turn roughly means that the input-output data dependence of `f_jvp` looks like the input-output dependence of `f_orig`.

But for `sinc` that's not the case: [the `_sinc_maclaurin` function has no data dependence on its input](https://github.com/google/jax/blob/6110be40dc66f0de9d0e3770645a7db81434910f/jax/_src/numpy/ufuncs.py#L703-L706), while [in its JVP rule the tangent output has a data dependence on the tangent input](https://github.com/google/jax/blob/6110be40dc66f0de9d0e3770645a7db81434910f/jax/_src/numpy/ufuncs.py#L711) (which it must, for linearity to hold). What resulted was ultimately a type error: we'd get a `custom_jvp_call_jaxpr` application with an `f32[]` output (and downstream operations, like `broadcast_in_dim`, which were set up for that `f32[]` output), but when differentiated we'd get an `f32[5]` primal (and `f32[5]` tangent), which was then type-incompatible with downstream applications.

There's a bit more going on here which was necessary to exhibit this bug; in particular, `scan` (i.e. `lax.map`) was necessary because it causes the JVP rule to be run in a later pass, rather than the JVP happening "on the fly" where we could've noticed the output of the JVP rule was batched. That's why replacing the `lax.map` application with a `jnp.stack([jnp.sinc(x_) for x_ in x])` did not exhibit the bug. (Sorry, this paragraph is probably even more inside-baseball than the preceding paragraph...)

This analysis directly leads to the two of the possible solutions briefly mentioned above:
1. if we had a `full_like` primitive then we could write `f` so as to model the same data dependence as `f_jvp` here;
2. if we defensively assume more stuff will come out batched in the `custom_jvp_call_jaxpr` vmap rule, we won't get this issue of JVPs looking batched where primals aren't.

I kind of like the first approach but I'm not sure. I might land the quick fix in the meantime, since multiplying by zero isn't so bad (and it encodes the data dependence we want to maintain).

**Comment by mattjj** [comment]
If we don't want to require users' `custom_jvp` functions to respect the data dependence invariant mentioned above, then we should go with solution 2, i.e. defensively batching more outputs. But that might have issues with `jacfwd` where we don't want primals to be batched but we do want tangents to be batched...

**Comment by mattjj** [comment]
Say we have

```python
f = jax.custom_jvp(f_orig)
f.defjvp(f_jvp)
```

A sufficient requirement for correctness here, which the original implementation of `sinc` violated (because `_sinc_maclaurin` was specializing its output to a particular input _point_) is that for all `x` we have

```python
f_jvp(x, t)[1] ≈ f_orig(x + t) - f_orig(x)
```

where `≈` means close as functions of `t` for all `t` in some region around 0. In particular, we want the derivatives to be the same at `t=0`.

This requirement is a property on the pair (`f_orig`, `f_jvp`).

But the original (`_sinc_maclaurin`, `_sinc_maclaurin_jvp`) violated this property because `_sinc_maclaurin` was a constant function while `lambda t: _sinc_maclaurin_jvp(x, t)[1]` was by necessity a linear function in `t`.

I like this statement of the requirement because it makes clear that we're only requiring that (`f_orig`, `f_jvp`) satisfy the defining characteristic of a JVP. But it's only sufficient and not necessary; the weaker necessary condition we require here is about data dependence, i.e. that `lambda t: f_jvp(x, t)[1]` not have any extra input-output data dependencies which `f_orig` does not have.

We should probably spell out this requirement on `custom_jvp` rules, and perhaps even have a cookbook recipe for a test. But it's quite an edge case: we only ran into it here because [`_sinc_maclaurin` is only applied when its input is zero (under a `jnp.where`)](https://github.com/google/jax/blob/6110be40dc66f0de9d0e3770645a7db81434910f/jax/_src/numpy/ufuncs.py#L695), leading us to write it as a constant function. But usually one doesn't write a `custom_jvp` of a constant function! Moreover, it's only an issue when combining both `vmap` and `scan` (or other control flow) on top.

**Comment by Ph03n1xdust** [comment]
Thank you for the quick fix, and also for the detailed explanation.
With those changes it works perfectly.

And thanks again for your work on jax!

**Comment by YouJiacheng** [comment]
Just curious, why `lax.full_like(x, constant)` doesn't model the data dependency correctly? Shouldn't it be equivalent to `x * 0 + constant`?

**Comment by mattjj** [comment]
@YouJiacheng great question, as always. As a Python function we would say `lax.full_like(x, constant)` has a data dependence on `x`. But as you know, JAX doesn't "see" Python functions directly: instead JAX _specializes_ Python functions while tracing. In this case, `jit` specializes on the shape and dtype of `x`, so these two are allindistinguishable from one another:

```python
x = jnp.ones(3, dtype='float32')

@jit
def f(x):
  return lax.full_like(x, 0)
f(x)

@jit
def g(x):
  return lax.full((3,), lax.convert_element_type(0, 'float32'))
g(x)
```

Notice that `g`, which is a shape-and-dtype-specialized version of `f`, has no data dependence on its input!

If instead `full_like` were a JAX primitive, then instead of trace-time-specializing away this Python data dependence, we would preserve the data dependence in the "JAX view" of things. That would give us a more natural way to express this particular code. (Indeed, because multiplication and addition are JAX primitives, that's why the `x * 0 + constant` spelling of the same function isn't dropping the data dependence.) But there would still be different spellings that would be problematic: `jnp.zeros(x.shape, x.dtype)` wouldn't be equivalent, for example.

I chalk this up to the complexity of JAX being a staged system: we're kind of programming and meta-programming at the same time, in that our Python expressions sometimes correspond fairly directly to jaxpr expressions but sometimes they act more like metaprogramming macros.

WDYT?

**Comment by YouJiacheng** [comment]
@mattjj Thanks for nice explanation, now I know why "I think the best fix may be to add a `full_like` primitive."
Since `vmap` operate on shapes, could it become easier to model data dependence(i.e. IIUC, preserve BatchTracer) by "propagating" tracer to shape for `BatchTracer.shape`? Hmmm.
```python
def f(x):
    y = lax.full_like(x, 0)
    print(x)
    print(y)
    return y
jax.vmap(f)(jnp.ones((3,3)))
# Traced<ShapedArray(float32[3])>with<BatchTrace(level=1/0)> with
#   val = DeviceArray([[1., 1., 1.],
#              [1., 1., 1.],
#              [1., 1., 1.]], dtype=float32)
#   batch_dim = 0
# [0. 0. 0.]
# expect y is
# Traced<ShapedArray(float32[3])>with<BatchTrace(level=1/0)> with
#   val = DeviceArray([[0., 0., 0.],
#              [0., 0., 0.],
#              [0., 0., 0.]], dtype=float32)
#   batch_dim = 0
# like y = x * 0 + 0

```

**Comment by mattjj** [comment]
Fixed by #10756!
